### PR TITLE
TimeDynamicGeoJson plugin 

### PR DESCRIPTION
**Pull request by halfdanrump**
In reference to [issue #722.](https://github.com/python-visualization/folium/issues/722)

This is a pull request to merge in a plugin called `TimeDynamicGeoJson`. 

- The main bulk of the code is in `folium/plugins/timedynamic_geo_json.py`. 
- The class uses d3. For now I put the script import in `folium/map.py`, but let me know if there's a better place to do this :). 
- There's a notebook demonstrating the plugin [here](https://github.com/halfdanrump/geoviz/blob/master/TimeSliderChoropleth.ipynb). I think it'd be useful to merge in that notebook as well, but it's probably better left for a later pull request. 

Cheers!

### CODE REVIEW

File: folium/map.py [hunk 1]
```diff
@@ -22,6 +22,10 @@
 
 ENV = Environment(loader=PackageLoader('folium', 'templates'))
 
+_d3_js = [
+    ('d3', 'http://d3js.org/d3.v4.min.js'),
```
Comment (ocefpaf) [inline folium/map.py]: This should be in the plugin and not in the global map.

File: folium/plugins/timedynamic_geo_json.py [hunk 2]
```diff
@@ -0,0 +1,257 @@
+import json
+from ..map import Layer
```
Comment (ocefpaf) [inline folium/plugins/timedynamic_geo_json.py]: Please use absolute imports to keep it consistent with the rest of the code.

File: folium/plugins/timedynamic_geo_json.py [hunk 3]
```diff
@@ -0,0 +1,257 @@
...
+from jinja2 import Template
+from branca.utilities import none_min, none_max, iter_points
```
Comment (ocefpaf) [inline folium/plugins/timedynamic_geo_json.py]: These are now in `folium.utilities`.

File: folium/plugins/timedynamic_geo_json.py [hunk 4]
```diff
@@ -0,0 +1,257 @@
...
+
+        if highlight_function is None:
+            def highlight_function(x):
+                return {}
+
+        self.highlight_function = highlight_function
+
+        self.smooth_factor = smooth_factor
+
+        self._template = Template(u"""
```
Comment (ocefpaf) [inline folium/plugins/timedynamic_geo_json.py]: You can probably inherit from the GeoJson class and save a lot of the duplication by only changing the template.

**Comment by ocefpaf** [comment]
> This is a pull request to merge in a plugin called `TimeDynamicGeoJson`. 

Looks cool! Thanks for the PR!!

> The class uses d3. For now I put the script import in `folium/map.py`, but let me know if there's a better place to do this :)

See https://github.com/python-visualization/folium/blob/master/folium/plugins/draw.py#L67-L71

> There's a notebook demonstrating the plugin [here](https://github.com/halfdanrump/geoviz/blob/master/TimeSliderChoropleth.ipynb). I think it'd be useful to merge in that notebook as well, but it's probably better left for a later pull request. 

Not really, you can add it to the examples folder in this PR. Remember to also add some tests and an entry to the changelog.

File: folium/plugins/timedynamic_geo_json.py (hunk 4 above)
Comment (halfdanrump) [reply folium/plugins/timedynamic_geo_json.py]: @ocefpaf Good idea!

File: folium/plugins/timedynamic_geo_json.py (hunk 3 above)
Comment (halfdanrump) [reply folium/plugins/timedynamic_geo_json.py]: Got it.

File: folium/plugins/timedynamic_geo_json.py (hunk 2 above)
Comment (halfdanrump) [reply folium/plugins/timedynamic_geo_json.py]: OK

File: folium/map.py (hunk 1 above)
Comment (halfdanrump) [reply folium/map.py]: Alright.

**Comment by halfdanrump** [comment]
@ocefpaf Thank you for reviewing the code! I'll implement the changes and push again :)

**Comment by halfdanrump** [comment]
@ocefpaf I've made the changes that you requested :). Buut:

- I keep getting linter errors. I use pylint on my local machine
- the test that passes locally fail on the CI. I'm trying to figure out why.

**Comment by ocefpaf** [comment]
> I keep getting linter errors. I use pylint on my local machine

Don't worry about the lints right now. We can fix those later.

> the test that passes locally fail on the CI. I'm trying to figure out why.

It seems that only Python 2.7 is failing. I'll download your code an investigate it locally.

File: folium/plugins/timedynamic_geo_json.py [hunk 5]
```diff
@@ -0,0 +1,175 @@
...
+    name : string, default None
+        The name of the Layer, as it will appear in LayerControls
+    overlay : bool, default False
+        Adds the layer as an optional overlay (True) or the base layer (False).
+    control : bool, default True
+        Whether the Layer will be included in LayerControls
+    smooth_factor: float, default None
+        How much to simplify the polyline on each zoom level. More means
+        better performance and smoother look, and less means more accurate
+        representation. Leaflet defaults to 1.0.
```
Comment (ocefpaf) [inline folium/plugins/timedynamic_geo_json.py]: You can probably remove all that, refer to the `GeoJson` docs, and add only what is different about this `plugin`.

File: folium/plugins/timedynamic_geo_json.py [hunk 6]
```diff
@@ -0,0 +1,175 @@
...
+                            ,
+                            {% endif %}
+                            onEachFeature: {{this.get_name()}}_onEachFeature
+                        {% endif %}
+                        }
+                    {% endif %}
+                    ).addTo({{this._parent.get_name()}}
+                );
+
+        	{{this.get_name()}}.setStyle(function(feature) {feature.properties.style;});
```
[the remark this answers is not in the document]
Comment (ocefpaf) [reply folium/plugins/timedynamic_geo_json.py]: Mix of space and tabs are a no-no. Can you try to find the tabs and make everything spaces?

File: folium/plugins/timedynamic_geo_json.py [hunk 7]
```diff
@@ -0,0 +1,175 @@
...
+
+            {% endmacro %}
+            """)
+
+    def render(self, **kwargs):
+        super(TimeDynamicGeoJson, self).render(**kwargs)
+        figure = self.get_root()
+        assert isinstance(figure, Figure), ('You cannot render this Element '
+                                            'if it is not in a Figure.')
+        figure.header.add_child(JavascriptLink('https://d3js.org/d3.v4.min.js'))  # noqa
```
Comment (ocefpaf) [inline folium/plugins/timedynamic_geo_json.py]: :+1:

File: folium/plugins/timedynamic_geo_json.py [hunk 8]
```diff
@@ -0,0 +1,175 @@
...
+                }
+
+                d3.select("#slider").on("input", function() {
+                    current_timestamp = timestamps[this.value];
+                var datestring = new Date(parseInt(current_timestamp)*1000).toDateString();
+                d3.select("output#slider-value").text(datestring);
+                fill_map();
+                });
+
+                {% if this.highlight %}
```
Comment (ocefpaf) [inline folium/plugins/timedynamic_geo_json.py]: Note to self: unfortunately this repetition is needed b/c of the current design we cannot extend the jinja template.

File: tests/plugins/test_timedynamic_geo_json.py [hunk 9]
```diff
@@ -0,0 +1,73 @@
...
+    out = m._parent.render()
+
+    # We verify that imports
+    assert '<script src="https://d3js.org/d3.v4.min.js' in out  # noqa
+
+    # We verify that data has been inserted currectly
+    expected_timestamps = ("var timestamps = ['1454166000', '1456671600', '1459350000', "
+                           "'1461942000', '1464620400', '1467212400', '1469890800', "
+                           "'1472569200', '1475161200', '1477839600'];")
+    #assert expected_timestamps in out
```
Comment (ocefpaf) [inline tests/plugins/test_timedynamic_geo_json.py]: Are you planning on un-commenting this part?

**Comment by ocefpaf** [comment]
I am preparing a PR to your fork with some changes that will help you finish this PR.
One quick question, how about we name this `TimeSliderChoropleth` instead of `TimeDynamicGeoJson`?

**Comment by halfdanrump** [comment]
@ocefpaf 

> I am preparing a PR to your fork with some changes that will help you finish this PR.

Thank you!

> One quick question, how about we name this TimeSliderChoropleth instead of TimeDynamicGeoJson?

It's a better name, I agree :).

**Comment by ocefpaf** [comment]
I was not able to send #741 to your branch, so I created a new PR. I'll try to finish the docs tomorrow and get it merged so you can polish the details if you want.

**Comment by halfdanrump** [comment]
> I was not able to send #741 to your branch, so I created a new PR. I'll try to finish the docs tomorrow and get it merged so you can polish the details if you want.

Alright, sounds good :))

**Comment by ocefpaf** [comment]
OK. I'll try to finalize this later this week. Thanks!

**Pull request merged**